frontend/packages/dev-console/src/components/project-access/ProjectAccessForm.tsx

```diff
@@ -0,0 +1,198 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { Formik } from 'formik';
+import { k8sCreate, k8sKill, k8sPatch } from '@console/internal/module/k8s';
+import { RoleBindingModel } from '@console/internal/models';
+import { LoadingBox, PageHeading } from '@console/internal/components/utils';
+import { generateSecret } from '../import/import-submit-utils';
+import { withStartGuide } from '../../../../../public/components/start-guide';
+import {
+  validationSchema,
+  UserRole,
+  UserRoleBinding,
+  getRolesWithNameChange,
+  getNewRoles,
+  getRemovedRoles,
+  getfinalRoles,
+} from './projectAccessForm-utils';
+
+export interface ProjectAccessFormProps {
+  ProjectAccessFormComponent: React.ComponentType<any>;
+  formName: string;
+  namespace: string;
+  roleBindings?: { data: []; loaded: boolean };
+  projects?: { data: []; loaded: boolean };
+}
+
+const ProjectAccessForm: React.FC<ProjectAccessFormProps> = ({
+  ProjectAccessFormComponent,
+  formName,
+  namespace,
+  roleBindings,
+  projects,
+}) => {
+  if (!projects.loaded || (projects.loaded && !_.isEmpty(projects.data) && !roleBindings.loaded)) {
+    return <LoadingBox />;
+  }
+
+  const noProjectsAvailable = projects.loaded && _.isEmpty(projects.data);
+
+  const filteredRoleBindings = _.filter(
+    roleBindings.data,
+    (user: UserRole) => user.roleRef.name === 'view' || user.roleRef.name === 'admin',
+  );
+  let userRoleBindings: UserRoleBinding[] = [];
+  filteredRoleBindings.map(
+    (user: UserRole) =>
+      (userRoleBindings = [
+        ...userRoleBindings,
+        ...[
+          {
+            roleBindingName: user.metadata.name,
+            user: user.subjects[0].name,
+            role: user.roleRef.name,
+          },
+        ],
+      ]),
+  );
+
+  const initialValues = {
+    projectAccess: roleBindings.loaded && userRoleBindings,
+  };
+
+  const generateRoleBindingName = (username: string, role: string) => {
```

Comment:
Lots of utility functions that are better put into a separate utility file for easier testing. Further to this, don't put functions within the render closure unless they make use of variables in the outer scope.

Reply:
Done